flax/linen/module.py

```diff
@@ -504,8 +504,9 @@ def __setattr__(self, name: str, val: Any):
       val: Value of the attribute.
     """
     is_dataclass_attr = name in self.__dataclass_fields__ and self.__dataclass_fields__[name].init  # pytype: disable=attribute-error
-    
-    if not self._state.in_setup and not is_dataclass_attr:
+    is_initialized = self.scope != None
```

Comment:
this will probably fail in setup and top level because a constructed module will not immediately get a scope.

Reply:
I've now added `is_initialized` to `Module` and added some more tests.